lua/nvim-treesitter/query.lua

```diff
@@ -46,8 +46,56 @@ function M.has_query_files(lang, query_name)
   return files and #files > 0
 end
 
-function M.get_query(lang, query_name)
-  return tsq.get_query(lang, query_name)
+do
+  local mt = {}
+  mt.__index = function(tbl, key)
+    if rawget(tbl, key) == nil then
+      rawset(tbl, key, {})
+    end
+    return rawget(tbl, key)
+  end
+
+  -- cache will have will auto set the table for each lang if it is nil
+  local cache = setmetatable({}, mt)
+
+  --- Same as `vim.treesitter.query` except will return cached values
+  function M.get_query(lang, query_name)
+    if cache[lang][query_name] == nil then
+      M.reload_file_cache(lang, query_name)
+    end
+
+    return cache[lang][query_name]
+  end
+
+  --- Reloads the query file cache.
+  --- If lang and query_name is both present, will reload for only the lang and query_name.
+  --- If only lang is present, will reload all query_names for that lang
+  --- If none are present, will reload everything
+  function M.reload_file_cache(lang, query_name)
+    -- nil means query is not cached yet. "invalid" means that query was cached and was not a valid one
+    if lang and query_name then
+      cache[lang][query_name] = tsq.get_query(lang, query_name)
+    elseif lang and not query_name then
+      for query_name, _ in pairs(cache[lang]) do
+        M.reload_file_cache(lang, query_name)
+      end
+    elseif not lang and not query_name then
+      for lang, _ in pairs(cache) do
+        for query_name, _ in pairs(cache[lang]) do
+          M.reload_file_cache(lang, query_name)
+        end
+      end
+    else
+      error("Cannot have query_name by itself!")
+    end
+  end
+end
+
+--- This function is meant for an autocommand and not to be used. Only use if file is a query file.
+function M.reload_file_cache_on_write(fname)
+  local fnamemodify = vim.fn.fnamemodify
+  print('reloading cache')
+  M.reload_file_cache(fnamemodify(fname, ':p:h:t'), fnamemodify(fname, ':t:r'))
 end
```

Comment:
Out of curiosity: why having that out of the `do ... end` block ?
This block really helps reading btw, so that's great !

Reply:
I put it out of the `do ... end` block for privacy so it cannot directly access the cache and instead has to call a function that will access the cache. That function pretty much only modifies the file so I thought that it should remain separate from the caching stuff.